File: FruitWholesale.Client/src/app/features/supply/supply.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';
import { PaginatedList, PaginationRequest } from '../../core/models/common.model';
import { SaveSupply, Supply, SupplyListItem } from '../../core/models/transactions.model';
import { toHttpParams } from '../../core/utils/http-params.util';

@Injectable({ providedIn: 'root' })
export class SupplyService {
  private readonly baseUrl = `${environment.apiUrl}/supply`;

  constructor(private readonly http: HttpClient) {}

  getPaged(request: PaginationRequest, shopId?: number | null, fromDate?: string | null, toDate?: string | null) {
    return this.http.get<PaginatedList<SupplyListItem>>(this.baseUrl, {
      params: toHttpParams({ ...request, shopId, fromDate, toDate })
    });
  }

  getById(id: number): Observable<Supply> {
    return this.http.get<Supply>(`${this.baseUrl}/${id}`);
  }

  getNextInvoiceNo(): Observable<string> {
    return this.http.get(`${this.baseUrl}/next-invoice-no`, { responseType: 'text' });
  }

  create(dto: SaveSupply): Observable<Supply> {
    return this.http.post<Supply>(this.baseUrl, dto);
  }

  update(id: number, dto: SaveSupply): Observable<Supply> {
    return this.http.put<Supply>(`${this.baseUrl}/${id}`, dto);
  }

  delete(id: number): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/${id}`);
  }
}

File: FruitWholesale.Mobile/lib/features/supply/supply_service.dart
import '../../core/api/api_client.dart';
import '../../core/models/paginated_list.dart';
import 'supply_models.dart';

class SupplyService {
  final ApiClient _api;

  SupplyService(this._api);

  Future<PaginatedList<SupplyListItem>> getPaged({
    int pageNumber = 1,
    int pageSize = 20,
    int? shopId,
    String? fromDate,
    String? toDate,
  }) async {
    final json = await _api.get('/supply', query: {
      'pageNumber': pageNumber,
      'pageSize': pageSize,
      if (shopId != null) 'shopId': shopId,
      if (fromDate != null) 'fromDate': fromDate,
      if (toDate != null) 'toDate': toDate,
    });
    return PaginatedList.fromJson(json as Map<String, dynamic>, SupplyListItem.fromJson);
  }

  Future<SupplyDetail> getById(int id) async {
    final json = await _api.get('/supply/$id');
    return SupplyDetail.fromJson(json as Map<String, dynamic>);
  }

  Future<SupplyBillExtras> getBillExtras(int id) async {
    final json = await _api.get('/supply/$id/bill');
    return SupplyBillExtras.fromJson(json as Map<String, dynamic>);
  }

  Future<String> getNextInvoiceNo() async {
    final json = await _api.get('/supply/next-invoice-no');
    return json as String;
  }

  Future<SupplyDetail> create(SupplyDetail supply) async {
    final json = await _api.post('/supply', body: supply.toSaveJson());
    return SupplyDetail.fromJson(json as Map<String, dynamic>);
  }

  Future<SupplyDetail> update(int id, SupplyDetail supply) async {
    final json = await _api.put('/supply/$id', body: supply.toSaveJson());
    return SupplyDetail.fromJson(json as Map<String, dynamic>);
  }
}
